### ISSUE TITLE
[Feature Request] seaborn.violinplot to handle two levels of hues

### ISSUE THREAD

**Issue by bheklilr** [mwaskom/seaborn#727]
Lately I've been running experiments that plot really well through `violinplot` as

    sns.violinplot(x='Description', y='Value', hue='Category', data=data)

where `Category` has more than 2 values, for this example I'll say there's 6 distinct categories and 4 distinct `Description`s, yielding a total of 24 violin plots on one subplot.  This is pretty ideal for a single experiment, but not when I want to compare two experiments to each other.  I can do

    experiment1['Experiment'] = 1
    experiment2['Experiment'] = 2
    both = pd.concat([experiment1, experiment2])

    sns.violinplot(x='Description', y='Value', hue='Experiment', data=both, split=True)

in order to have a comparison between the two experiments, but now I lose the granularity of having `hue='Category'`.  Currently I have to do a lot of work to get them all on the same subplot, or I can take the lazy route of having separate multiple subplots, one for each category.  Ideally what I would like is to have

    sns.violinplot(x='Description', y='Value', hue=['Category', 'Experiment'], data=both, split=True)

And end up with 24 split violins, 6 for each description where each set of 6 are split by the experiment.

There are several concerns I have with implementing this:

* This only works with 2 hues where the second one can be split.  This is really just an extension of the current limitation on splitting violins.
* This would only work where the second hue is always split, so that would have to override whatever value is passed in for `split`.
* How should they be colored or filled in?  Having the background color set by the first hue and then having hatches for the second hue might work.
* Would it make more sense to have `split='Experiment'` compared to `hue=['Category', 'Experiment'], split=True`?  This would then suggest to users that `hue` and `split` perform similar duties and a call like `sns.violinplot(x='Description', y='Value', split='Experiment', ...)` would be valid without specifying `hue`.
* Would this make more sense to implement this as a separate function to avoid confusion on the part of the user?

Does this sound like a worthwhile feature to add?  I certainly would get a lot of mileage out of it, but it seems to be a pretty specific use case.

**Comment by mwaskom** [comment]
I don't think this is going to work, but you could use facets to get additional levels of nesting.

**Comment by bheklilr** [comment]
@mwaskom 
Thanks for your quick response!  FacetGrid certainly helps set up the different subplots, but it's still not quite the view I'd like.  Again, I understand this is a pretty specific use case and it isn't terribly likely to get implemented.  At the same time, it seems like a natural extension of being able to spread out by a category and split based on another category, since that's basically the equivalent of having more x-axis values with splitting.  The biggest difference there is spacing, x-axis labels, and automatically handling the fill colors/hatching.  What technical difficulties do you see getting in the way of implementing this?  I wouldn't mind trying to implement it myself (I've been digging down into the seaborn source this week to figure out why some of my attempts at plotting were raising errors).  After reading the issue submission guidelines I saw that it was best to open an issue for discussion first to find out if it's even feasible.

**Comment by shoyer** [comment]
Some examples would be helpful here in explaining exactly what you're looking for.

In my experience, using a FacetGrid works pretty well for these use cases.

**Comment by mwaskom** [comment]
> What technical difficulties do you see getting in the way of implementing this?

Just from knowing how the code is written, it's not set up to add arbitrary levels of nesting. Also the violinplot is just one interface into the categorical plotting code that is shared with barplot, pointplot, boxplot, etc. If you wanted to add additional levels of nesting, you'd have to find some way to make them work for those kinds of plots too.

**Comment by bheklilr** [comment]
@shoyer [Here's](https://gist.github.com/bheklilr/38d0e66214564089d61f) a gist with some example data and the code to do the kind of plot I'm looking for, but using `FacetGrid` (the code snippet is at the bottom of the page), and [here's](http://i.imgur.com/E74zlYT.png) the image that I generated using that code.  You may have to run `tight_layout()` and the `subplots_adjust(wspace=0)` again after maximizing the figure to get a decent looking plot, so ipython will be useful.

Basically this graph looks pretty much exactly how I want, but I would prefer if it were 1 subplot instead of 6.  I'm having to fake that right now by setting `wspace=0`.  If it were a single subplot then I could zoom and pan across the entire graph instead of just one Description at a time. Also, I normally would plot more lines on top of this to represent limits that the violins should be within.  Having just one subplot to worry about would make this much easier to work with.

**Comment by bheklilr** [comment]
@mwaskom I wouldn't necessarily want arbitrary levels of nesting, just one additional level.  Knowing that the violinplot shares code with several other kinds of charts does complicate things, I hadn't considered that it would affect those plots so much, too.  Still, if I have time this weekend I'll take a crack at it.  Thanks for you input :)

**Comment by shoyer** [comment]
@bheklilr if you want to do things like plot extra lines on each subplot, that's exactly what `FacetGrid.map` is for, e.g., `fg.map(plt.axhline)`. My two cents is that this isn't a great idea because of the added complexity.

**Comment by bheklilr** [comment]
@shoyer Unfortunately `map` wouldn't work in this case because I would want different arguments to `axhline` on each subplot (and I haven't figured out how to do that yet, if there is a way), but it does sound like a pretty complicated feature to add as of right now that would make all of those functions harder to use.  Thanks again for your advice.

**Comment by MonsieurWave** [comment]
Same issue here for `sns.boxplot`. 
Any other possibilities for this? 
For exampled coloring style to add another level of data? One could imaging that some datapoints would be fully coloured, others striped...
